File: modules/vesc-ble/android/src/main/java/expo/modules/vescble/RefloatConfigEncoder.kt
package expo.modules.vescble

import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.roundToInt

internal class RefloatConfigEncodeException(message: String) : Exception(message)

// @parity /modules/vesc-ble/ios/RefloatConfigEncoder.swift
internal object RefloatConfigEncoder {
  fun encode(
    schema: RefloatConfigSchema,
    rawConfig: ByteArray,
    fields: Map<String, Any>,
  ): ByteArray {
    val result = rawConfig.copyOf()
    val byId = schema.fields.associateBy { it.id }
    for ((fieldId, value) in fields) {
      val schemaField = byId[fieldId] ?: continue
      writeValue(result, schemaField, value)
    }
    return result
  }

  private fun writeValue(bytes: ByteArray, field: RefloatConfigSchemaField, value: Any) {
    val view = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)
    view.position(field.offset)
    when (field.type) {
      RefloatConfigValueType.FLOAT32 -> view.putFloat(toDouble(value).toFloat())
      RefloatConfigValueType.FLOAT32_SCALED -> {
        val scale = requireScale(field)
        view.putInt((toDouble(value) * scale).roundToInt())
      }
      RefloatConfigValueType.FLOAT32_AUTO -> writeFloat32Auto(bytes, field.offset, toDouble(value))
      RefloatConfigValueType.FLOAT16_SCALED -> {
        val scale = requireScale(field)
        view.putShort((toDouble(value) * scale).roundToInt().toShort())
      }
      RefloatConfigValueType.INT32 -> view.putInt(toDouble(value).toInt())
      RefloatConfigValueType.UINT32 -> view.putInt(toDouble(value).toLong().toInt())
      RefloatConfigValueType.INT16 -> view.putShort(toDouble(value).toInt().toShort())
      RefloatConfigValueType.UINT16 -> view.putShort(toDouble(value).toInt().toShort())
      RefloatConfigValueType.INT8 -> view.put(toDouble(value).toInt().toByte())
      RefloatConfigValueType.UINT8 -> view.put(toDouble(value).toInt().toByte())
      RefloatConfigValueType.BOOL -> view.put(if (toBool(value)) 1.toByte() else 0.toByte())
    }
  }

  private fun writeFloat32Auto(bytes: ByteArray, offset: Int, value: Double) {
    if (value == 0.0) {
      ByteBuffer.wrap(bytes, offset, 4).order(ByteOrder.BIG_ENDIAN).putInt(0)
      return
    }
    val neg = value < 0.0
    val absVal = abs(value)
    val e = floor(ln(absVal) / ln(2.0)).toInt() + 126
    val eRaw = e.coerceIn(0, 255)
    val sig = absVal / Math.pow(2.0, (eRaw - 126).toDouble()) - 0.5
    val sigI = (sig * 8388608.0 * 2.0).roundToInt().coerceIn(0, 0x7fffff)
    var raw = (eRaw shl 23) or sigI
    if (neg) raw = raw or (1 shl 31)
    ByteBuffer.wrap(bytes, offset, 4).order(ByteOrder.BIG_ENDIAN).putInt(raw)
  }

  private fun requireScale(field: RefloatConfigSchemaField): Double {
    return field.scale
      ?: throw RefloatConfigEncodeException("CONFIG_ENCODE_FAILED: missing scale for ${field.id}")
  }

  private fun toDouble(value: Any): Double = when (value) {
    is Double -> value
    is Float -> value.toDouble()
    is Int -> value.toDouble()
    is Long -> value.toDouble()
    is Number -> value.toDouble()
    else -> throw RefloatConfigEncodeException("CONFIG_ENCODE_FAILED: cannot convert $value to Double")
  }

  private fun toBool(value: Any): Boolean = when (value) {
    is Boolean -> value
    is Number -> value.toInt() != 0
    else -> throw RefloatConfigEncodeException("CONFIG_ENCODE_FAILED: cannot convert $value to Boolean")
  }
}

File: modules/vescape-core/ios/config/RefloatConfigEncoder.swift
import Foundation

struct RefloatConfigEncodeException: Error {
  let message: String
  init(_ message: String) { self.message = message }
}

/// Patches Refloat config bytes with Tune Profile field values, big-endian, byte-for-byte identical
/// to the Android encoder so a write verifies against the same board readback.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/config/RefloatConfigEncoder.kt
enum RefloatConfigEncoder {
  static func encode(
    schema: RefloatConfigSchema,
    rawConfig: [UInt8],
    fields: [String: Any]
  ) throws -> [UInt8] {
    var result = rawConfig
    let byId = Dictionary(uniqueKeysWithValues: schema.fields.map { ($0.id, $0) })
    for (fieldId, value) in fields {
      guard let field = byId[fieldId] else { continue }
      try writeValue(&result, field, value)
    }
    return result
  }

  private static func writeValue(_ bytes: inout [UInt8], _ field: RefloatConfigSchemaField, _ value: Any) throws {
    switch field.type {
    case .float32:
      putUInt32(&bytes, field.offset, Float(try toDouble(value)).bitPattern)
    case .float32Scaled:
      let scale = try requireScale(field)
      putUInt32(&bytes, field.offset, UInt32(bitPattern: int32Saturating(roundHalfUp(try toDouble(value) * scale))))
    case .float32Auto:
      writeFloat32Auto(&bytes, field.offset, try toDouble(value))
    case .float16Scaled:
      let scale = try requireScale(field)
      putUInt16(&bytes, field.offset, UInt16(truncatingIfNeeded: int32Saturating(roundHalfUp(try toDouble(value) * scale))))
    case .int32:
      putUInt32(&bytes, field.offset, UInt32(bitPattern: int32Saturating(try toDouble(value).rounded(.towardZero))))
    case .uint32:
      putUInt32(&bytes, field.offset, UInt32(truncatingIfNeeded: int64Saturating(try toDouble(value).rounded(.towardZero))))
    case .int16:
      putUInt16(&bytes, field.offset, UInt16(truncatingIfNeeded: int32Saturating(try toDouble(value).rounded(.towardZero))))
    case .uint16:
      putUInt16(&bytes, field.offset, UInt16(truncatingIfNeeded: int32Saturating(try toDouble(value).rounded(.towardZero))))
    case .int8:
      bytes[field.offset] = UInt8(truncatingIfNeeded: int32Saturating(try toDouble(value).rounded(.towardZero)))
    case .uint8:
      bytes[field.offset] = UInt8(truncatingIfNeeded: int32Saturating(try toDouble(value).rounded(.towardZero)))
    case .bool:
      bytes[field.offset] = try toBool(value) ? 1 : 0
    }
  }

  private static func writeFloat32Auto(_ bytes: inout [UInt8], _ offset: Int, _ value: Double) {
    if value == 0.0 {
      putUInt32(&bytes, offset, 0)
      return
    }
    let neg = value < 0.0
    let absVal = abs(value)
    let e = Int((log(absVal) / log(2.0)).rounded(.down)) + 126
    let eRaw = min(max(e, 0), 255)
    let sig = absVal / pow(2.0, Double(eRaw - 126)) - 0.5
    let sigI = min(max(Int(roundHalfUp(sig * 8_388_608.0 * 2.0)), 0), 0x7fffff)
    var raw = (UInt32(eRaw) << 23) | UInt32(sigI)
    if neg { raw |= (UInt32(1) << 31) }
    putUInt32(&bytes, offset, raw)
  }

  private static func requireScale(_ field: RefloatConfigSchemaField) throws -> Double {
    guard let scale = field.scale else {
      throw RefloatConfigEncodeException("CONFIG_ENCODE_FAILED: missing scale for \(field.id)")
    }
    return scale
  }

  private static func toDouble(_ value: Any) throws -> Double {
    switch value {
    case let v as Double: return v
    case let v as Float: return Double(v)
    case let v as Int: return Double(v)
    case let v as Int64: return Double(v)
    case let v as Bool: return v ? 1.0 : 0.0
    case let v as NSNumber: return v.doubleValue
    default:
      throw RefloatConfigEncodeException("CONFIG_ENCODE_FAILED: cannot convert \(value) to Double")
    }
  }

  private static func toBool(_ value: Any) throws -> Bool {
    switch value {
    case let v as Bool: return v
    case let v as Int: return v != 0
    case let v as Double: return v != 0.0
    case let v as NSNumber: return v.intValue != 0
    default:
      throw RefloatConfigEncodeException("CONFIG_ENCODE_FAILED: cannot convert \(value) to Boolean")
    }
  }

  /// Kotlin `roundToInt` rounds half toward positive infinity (`floor(x + 0.5)`); match it so scaled
  /// and auto-float fields serialize to identical bytes.
  private static func roundHalfUp(_ value: Double) -> Double { (value + 0.5).rounded(.down) }

  private static func int32Saturating(_ value: Double) -> Int32 {
    if value >= Double(Int32.max) { return Int32.max }
    if value <= Double(Int32.min) { return Int32.min }
    return Int32(value)
  }

  private static func int64Saturating(_ value: Double) -> Int64 {
    if value >= Double(Int64.max) { return Int64.max }
    if value <= Double(Int64.min) { return Int64.min }
    return Int64(value)
  }

  private static func putUInt32(_ bytes: inout [UInt8], _ offset: Int, _ value: UInt32) {
    bytes[offset] = UInt8((value >> 24) & 0xff)
    bytes[offset + 1] = UInt8((value >> 16) & 0xff)
    bytes[offset + 2] = UInt8((value >> 8) & 0xff)
    bytes[offset + 3] = UInt8(value & 0xff)
  }

  private static func putUInt16(_ bytes: inout [UInt8], _ offset: Int, _ value: UInt16) {
    bytes[offset] = UInt8((value >> 8) & 0xff)
    bytes[offset + 1] = UInt8(value & 0xff)
  }
}
